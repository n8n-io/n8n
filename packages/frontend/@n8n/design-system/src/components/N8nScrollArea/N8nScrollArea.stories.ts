import type { StoryFn } from '@storybook/vue3';

import N8nScrollArea from './N8nScrollArea.vue';

export default {
	title: 'Atoms/ScrollArea',
	component: N8nScrollArea,
	argTypes: {
		type: {
			control: 'select',
			options: ['auto', 'always', 'scroll', 'hover'],
		},
		dir: {
			control: 'select',
			options: ['ltr', 'rtl'],
		},
		scrollHideDelay: {
			control: 'number',
		},
		maxHeight: {
			control: 'text',
		},
		maxWidth: {
			control: 'text',
		},
		enableHorizontalScroll: {
			control: 'boolean',
		},
		enableVerticalScroll: {
			control: 'boolean',
		},
	},
};

const Template: StoryFn = (args) => ({
	setup() {
		return { args };
	},
	components: {
		N8nScrollArea,
	},
	template: `
		<div style="width: 300px; height: 200px; border: 1px solid var(--color-foreground-base); border-radius: 4px;">
			<N8nScrollArea v-bind="args">
				<div style="padding: 16px;">
					<h3 style="margin: 0 0 12px 0; font-size: 16px; font-weight: 600;">Scrollable Content</h3>
					<p style="margin: 0 0 12px 0; line-height: 1.6;">
						This is a scrollable area with custom styled scrollbars. The content will scroll when it overflows the container.
					</p>
					<p style="margin: 0 0 12px 0; line-height: 1.6;">
						Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
					</p>
					<p style="margin: 0 0 12px 0; line-height: 1.6;">
						Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
					</p>
					<p style="margin: 0 0 12px 0; line-height: 1.6;">
						Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
					</p>
					<p style="margin: 0 0 12px 0; line-height: 1.6;">
						Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
					</p>
					<p style="margin: 0; line-height: 1.6;">
						Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.
					</p>
				</div>
			</N8nScrollArea>
		</div>
	`,
});

export const Default = Template.bind({});
Default.args = {
	type: 'hover',
	enableVerticalScroll: true,
	enableHorizontalScroll: false,
};

export const AlwaysVisible = Template.bind({});
AlwaysVisible.args = {
	type: 'always',
	enableVerticalScroll: true,
	enableHorizontalScroll: false,
};

export const WithMaxHeight = Template.bind({});
WithMaxHeight.args = {
	type: 'hover',
	maxHeight: '150px',
	enableVerticalScroll: true,
	enableHorizontalScroll: false,
};

const HorizontalScrollTemplate: StoryFn = (args) => ({
	setup() {
		return { args };
	},
	components: {
		N8nScrollArea,
	},
	template: `
		<div style="width: 300px; height: 100px; border: 1px solid var(--color-foreground-base); border-radius: 4px;">
			<N8nScrollArea v-bind="args">
				<div style="padding: 16px; white-space: nowrap; width: 600px;">
					<span style="font-weight: 600;">Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.

Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.

Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.

Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.

Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.</span>
					<div style="margin-left: 12px; max-width: 200px; text-wrap: auto">Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.

Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.

Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.

Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.

Lorem ipsum dolor sit amet consectetur adipiscing elit. Quisque faucibus ex sapien vitae pellentesque sem placerat. In id cursus mi pretium tellus duis convallis. Tempus leo eu aenean sed diam urna tempor. Pulvinar vivamus fringilla lacus nec metus bibendum egestas. Iaculis massa nisl malesuada lacinia integer nunc posuere. Ut hendrerit semper vel class aptent taciti sociosqu. Ad litora torquent per conubia nostra inceptos himenaeos.</div>
				</div>
			</N8nScrollArea>
		</div>
	`,
});

export const HorizontalScroll = HorizontalScrollTemplate.bind({});
HorizontalScroll.args = {
	type: 'hover',
	enableVerticalScroll: false,
	enableHorizontalScroll: true,
};

export const BothDirections = HorizontalScrollTemplate.bind({});
BothDirections.args = {
	type: 'hover',
	enableVerticalScroll: true,
	enableHorizontalScroll: true,
};

const InPopoverTemplate: StoryFn = (args) => ({
	setup() {
		return { args };
	},
	components: {
		N8nScrollArea,
	},
	template: `
		<div style="width: 260px; padding: 16px; background-color: var(--color-foreground-xlight); border: var(--border-base); border-radius: var(--border-radius-base); box-shadow: rgba(0, 0, 0, 0.1) 0px 10px 15px -3px, rgba(0, 0, 0, 0.05) 0px 4px 6px -2px;">
			<N8nScrollArea v-bind="args">
				<div style="display: flex; flex-direction: column; gap: 12px;">
					<h3 style="margin: 0; font-size: 14px; font-weight: 600;">Long Menu Items</h3>
					<div v-for="i in 15" :key="i" style="padding: 8px 12px; background: var(--color-background-base); border-radius: 4px; cursor: pointer;">
						Menu item {{ i }}: Some descriptive text that might be quite long
					</div>
				</div>
			</N8nScrollArea>
		</div>
	`,
});

export const InPopover = InPopoverTemplate.bind({});
InPopover.args = {
	type: 'hover',
	maxHeight: '200px',
	enableVerticalScroll: true,
	enableHorizontalScroll: false,
};
InPopover.storyName = 'Popover Example';
