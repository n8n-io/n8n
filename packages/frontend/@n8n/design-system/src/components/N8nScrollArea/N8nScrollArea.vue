<script setup lang="ts">
import {
	ScrollAreaCorner,
	ScrollAreaRoot,
	ScrollAreaScrollbar,
	ScrollAreaThumb,
	ScrollAreaViewport,
} from 'reka-ui';
import { computed } from 'vue';

export interface Props {
	/**
	 * Controls scrollbar visibility behavior
	 * - auto: shows scrollbars only when content overflows
	 * - always: always shows scrollbars
	 * - scroll: shows scrollbars when scrolling
	 * - hover: shows scrollbars on hover
	 */
	type?: 'auto' | 'always' | 'scroll' | 'hover';
	/**
	 * Reading direction for RTL support
	 */
	dir?: 'ltr' | 'rtl';
	/**
	 * Time in milliseconds before scrollbars auto-hide
	 */
	scrollHideDelay?: number;
	/**
	 * Maximum height of the scroll area
	 */
	maxHeight?: string;
	/**
	 * Maximum width of the scroll area
	 */
	maxWidth?: string;
	/**
	 * Whether to show horizontal scrollbar
	 */
	enableHorizontalScroll?: boolean;
	/**
	 * Whether to show vertical scrollbar
	 */
	enableVerticalScroll?: boolean;
}

const props = withDefaults(defineProps<Props>(), {
	type: 'hover',
	dir: 'ltr',
	scrollHideDelay: 600,
	maxHeight: undefined,
	maxWidth: undefined,
	enableHorizontalScroll: false,
	enableVerticalScroll: true,
});

const viewportStyle = computed(() => {
	const style: Record<string, string> = {};
	if (props.maxHeight) {
		style.maxHeight = props.maxHeight;
	}
	if (props.maxWidth) {
		style.maxWidth = props.maxWidth;
	}
	return style;
});
</script>

<template>
	<ScrollAreaRoot
		:type="type"
		:dir="dir"
		:scroll-hide-delay="scrollHideDelay"
		:class="$style.scrollAreaRoot"
	>
		<ScrollAreaViewport :class="$style.viewport" :style="viewportStyle">
			<slot />
		</ScrollAreaViewport>

		<ScrollAreaScrollbar
			v-if="enableVerticalScroll"
			orientation="vertical"
			:class="$style.scrollbar"
		>
			<ScrollAreaThumb :class="$style.thumb" />
		</ScrollAreaScrollbar>

		<ScrollAreaScrollbar
			v-if="enableHorizontalScroll"
			orientation="horizontal"
			:class="$style.scrollbar"
		>
			<ScrollAreaThumb :class="$style.thumb" />
		</ScrollAreaScrollbar>

		<ScrollAreaCorner v-if="enableHorizontalScroll && enableVerticalScroll" />
	</ScrollAreaRoot>
</template>

<style lang="scss" module>
.scrollAreaRoot {
	position: relative;
	overflow: hidden;
	width: 100%;
	height: 100%;
	--scrollbar-size: 10px;
}

.viewport {
	width: 100%;
	height: 100%;
	border-radius: inherit;
}

.scrollbar {
	display: flex;
	user-select: none;
	touch-action: none;
	padding: var(--spacing-5xs);
	background: transparent;
	transition: background 160ms ease-out;

	&:hover {
		background: var(--color-foreground-light);
	}

	&[data-orientation='vertical'] {
		width: var(--spacing-xs);
	}

	&[data-orientation='horizontal'] {
		height: var(--spacing-xs);
		flex-direction: row;
	}
}

.thumb {
	flex: 1;
	background: var(--color-foreground-base);
	border-radius: 4px;
	position: relative;

	&::before {
		content: '';
		position: absolute;
		top: 50%;
		left: 50%;
		transform: translate(-50%, -50%);
		width: 100%;
		height: 100%;
		min-width: 44px;
		min-height: 44px;
	}

	&:hover {
		background: var(--color-foreground-dark);
	}

	&:active {
		background: var(--color-foreground-xdark);
	}
}

// Style the scrollbar when type is 'always' to be more subtle
.scrollAreaRoot[data-type='always'] {
	.scrollbar {
		background: var(--color-foreground-xlight);

		&:hover {
			background: var(--color-foreground-light);
		}
	}

	.thumb {
		background: var(--color-foreground-light);

		&:hover {
			background: var(--color-foreground-base);
		}
	}
}

// Enhanced styling for hover type
.scrollAreaRoot[data-type='hover'] {
	.scrollbar {
		opacity: 0;
		transition: opacity 160ms ease-out;
	}

	&:hover .scrollbar {
		opacity: 1;
	}
}
</style>
