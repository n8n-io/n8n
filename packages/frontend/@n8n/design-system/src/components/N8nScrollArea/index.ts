import N8nScrollArea from './N8nScrollArea.vue';

export default N8nScrollArea;
