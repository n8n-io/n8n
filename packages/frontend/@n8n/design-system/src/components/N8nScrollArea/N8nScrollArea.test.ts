import { render } from '@testing-library/vue';

import N8nScrollArea from './N8nScrollArea.vue';

describe('N8nScrollArea', () => {
	it('should render correctly with default props', () => {
		const wrapper = render(N8nScrollArea, {
			slots: {
				default: '<div>Test content</div>',
			},
		});

		expect(wrapper.container).toMatchSnapshot();
	});

	it('should render with custom type prop', () => {
		const wrapper = render(N8nScrollArea, {
			props: {
				type: 'always',
			},
			slots: {
				default: '<div>Test content</div>',
			},
		});

		const scrollAreaRoot = wrapper.container.querySelector('.scrollAreaRoot');
		expect(scrollAreaRoot).toBeInTheDocument();
	});

	it('should render with maxHeight style when provided', () => {
		const wrapper = render(N8nScrollArea, {
			props: {
				maxHeight: '200px',
			},
			slots: {
				default: '<div>Test content</div>',
			},
		});

		const viewport = wrapper.container.querySelector('[style*="max-height"]');
		expect(viewport).toHaveStyle({ maxHeight: '200px' });
	});

	it('should render with maxWidth style when provided', () => {
		const wrapper = render(N8nScrollArea, {
			props: {
				maxWidth: '300px',
			},
			slots: {
				default: '<div>Test content</div>',
			},
		});

		const viewport = wrapper.container.querySelector('[style*="max-width"]');
		expect(viewport).toHaveStyle({ maxWidth: '300px' });
	});

	it('should render with enableVerticalScroll prop', () => {
		const wrapper = render(N8nScrollArea, {
			props: {
				enableVerticalScroll: true,
			},
			slots: {
				default: '<div>Test content</div>',
			},
		});

		// Check that the component renders successfully with the prop
		expect(wrapper.container.querySelector('.scrollAreaRoot')).toBeInTheDocument();
		expect(wrapper.container.querySelector('.viewport')).toBeInTheDocument();
	});

	it('should render with enableHorizontalScroll prop', () => {
		const wrapper = render(N8nScrollArea, {
			props: {
				enableHorizontalScroll: true,
			},
			slots: {
				default: '<div>Test content</div>',
			},
		});

		// Check that the component renders successfully with the prop
		expect(wrapper.container.querySelector('.scrollAreaRoot')).toBeInTheDocument();
		expect(wrapper.container.querySelector('.viewport')).toBeInTheDocument();
	});

	it('should render with both scroll directions enabled', () => {
		const wrapper = render(N8nScrollArea, {
			props: {
				enableVerticalScroll: true,
				enableHorizontalScroll: true,
			},
			slots: {
				default: '<div>Test content</div>',
			},
		});

		// Check that the component renders successfully with both props
		expect(wrapper.container.querySelector('.scrollAreaRoot')).toBeInTheDocument();
		expect(wrapper.container.querySelector('.viewport')).toBeInTheDocument();
	});

	it('should render with scroll disabled', () => {
		const wrapper = render(N8nScrollArea, {
			props: {
				enableVerticalScroll: false,
				enableHorizontalScroll: false,
			},
			slots: {
				default: '<div>Test content</div>',
			},
		});

		// Check that the component still renders properly even with scrollbars disabled
		expect(wrapper.container.querySelector('.scrollAreaRoot')).toBeInTheDocument();
		expect(wrapper.container.querySelector('.viewport')).toBeInTheDocument();
	});

	it('should pass dir prop to ScrollAreaRoot', () => {
		const wrapper = render(N8nScrollArea, {
			props: {
				dir: 'rtl',
			},
			slots: {
				default: '<div>Test content</div>',
			},
		});

		const scrollAreaRoot = wrapper.container.querySelector('.scrollAreaRoot');
		expect(scrollAreaRoot).toHaveAttribute('dir', 'rtl');
	});

	it('should pass scrollHideDelay prop to ScrollAreaRoot', () => {
		const wrapper = render(N8nScrollArea, {
			props: {
				scrollHideDelay: 1000,
			},
			slots: {
				default: '<div>Test content</div>',
			},
		});

		// Note: This would need to be tested with more sophisticated testing
		// as the scrollHideDelay is an internal prop that affects behavior
		expect(wrapper.container).toMatchSnapshot();
	});
});
