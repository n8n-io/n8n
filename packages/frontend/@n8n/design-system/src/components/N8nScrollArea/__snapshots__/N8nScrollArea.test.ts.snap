// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`N8nScrollArea > should pass scrollHideDelay prop to ScrollAreaRoot 1`] = `
<div>
  <div
    class="scrollAreaRoot"
    dir="ltr"
    style="position: relative; --reka-scroll-area-corner-width: 0px; --reka-scroll-area-corner-height: 0px;"
  >
    
    
    <div
      class="viewport"
      data-reka-scroll-area-viewport=""
      style="overflow-x: hidden; overflow-y: hidden;"
      tabindex="0"
    >
      <div>
        
        
        <div>
          Test content
        </div>
        
        
      </div>
    </div>
    <style>
       /* Hide scrollbars cross-browser and enable momentum scroll for touch devices */ [data-reka-scroll-area-viewport] { scrollbar-width:none; -ms-overflow-style:none; -webkit-overflow-scrolling:touch; } [data-reka-scroll-area-viewport]::-webkit-scrollbar { display:none; } 
    </style>
    
    <!---->
    <!--v-if-->
    <!--v-if-->
    
  </div>
</div>
`;

exports[`N8nScrollArea > should render correctly with default props 1`] = `
<div>
  <div
    class="scrollAreaRoot"
    dir="ltr"
    style="position: relative; --reka-scroll-area-corner-width: 0px; --reka-scroll-area-corner-height: 0px;"
  >
    
    
    <div
      class="viewport"
      data-reka-scroll-area-viewport=""
      style="overflow-x: hidden; overflow-y: hidden;"
      tabindex="0"
    >
      <div>
        
        
        <div>
          Test content
        </div>
        
        
      </div>
    </div>
    <style>
       /* Hide scrollbars cross-browser and enable momentum scroll for touch devices */ [data-reka-scroll-area-viewport] { scrollbar-width:none; -ms-overflow-style:none; -webkit-overflow-scrolling:touch; } [data-reka-scroll-area-viewport]::-webkit-scrollbar { display:none; } 
    </style>
    
    <!---->
    <!--v-if-->
    <!--v-if-->
    
  </div>
</div>
`;
